### ISSUE TITLE
Number of IO Completion and Submission Queues Allocated defaulted to 1

### ISSUE THREAD

**Issue by sridharbalachandriah** [spdk/spdk#3111]
# Sighting report

while evaluating spdk-bdev NVMF over TCP transport .I am seeing a skew in the queue settings, where it shows the number of queues  as 1.

nvme get-feature /dev/nvme0n1 -f 7 -H
get-feature:0x07 (Number of Queues), Current value:00000000
        Number of IO Completion Queues Allocated (NCQA): 1
        Number of IO Submission Queues Allocated (NSQA): 1

## Expected Behavior

 nvme get-feature /dev/nvme1 -f 7 -H
get-feature:0x7 (Number of Queues), Current value:0x7e007e
        Number of IO Completion Queues Allocated (NCQA): 127
        Number of IO Submission Queues Allocated (NSQA): 127

## Possible Solution

I have tried manually setting QPairs per controller, along with with Queue depth, but the values turn up as 1
./scripts/rpc.py nvmf_create_transport -t tcp -u 131072 -i 131072  -q 128 -m 128

What are default values? Is there a spdk command or a workaround to set the NCQA and NSQA?

## Steps to Reproduce

1.
2.
3.
4.

## Context (Environment including OS version, SPDK version, etc.)
./build/bin/nvmf_tgt -v
SPDK v23.09-pre git sha1 9c6458d98

OS version
LSB Version:    :core-4.1-amd64:core-4.1-noarch
Distributor ID: Rocky
Description:    Rocky Linux release 8.7 (Green Obsidian)
Release:        8.7

**Comment by ksztyber** [comment]
Hi @sridharbalachandriah, it shows up as 127 (which is the default in SPDK) on my system. And SPDK always returns that value as ncqa/nsqa, regardless of what the initiator asks for, so I think this is most likely caused by the kernel driver, especially since you're running a pretty old kernel. Could you recheck on something newer?

Also, could you post the value of:
```
scripts/rpc.py nvmf_get_transports
```
and
```
nvme admin-passthru -o 0xa -4 7 /dev/nvme0
```

**Comment by sridharbalachandriah** [comment]
Hi @ksztyber ,

 scripts/rpc.py nvmf_get_transports
[
  {
    "trtype": "TCP",
    "max_queue_depth": 128,
    "max_io_qpairs_per_ctrlr": 127,
    "in_capsule_data_size": 4096,
    "max_io_size": 131072,
    "io_unit_size": 131072,
    "max_aq_depth": 128,
    "num_shared_buffers": 511,
    "buf_cache_size": 4294967295,
    "dif_insert_or_strip": false,
    "zcopy": false,
    "c2h_success": true,
    "sock_priority": 0,
    "abort_timeout_sec": 1
  }
]

nvme admin-passthru -o 0xa -4 7 /dev/nvme0
Admin Command Get Features is Success and result: 0x00000000

1. Any suggested version of the kernel driver?

**Comment by ksztyber** [comment]
Could you also post the output of `build/examples/identify`?

> Any suggested version of the kernel driver?

I've tested it on 6.1, it's the default on Fedora 37.

**Comment by sridharbalachandriah** [comment]
# LD_LIBRARY_PATH=/opt/daos/lib64 ./build/examples/identify -r 'trtype:TCP adrfam:IPv4 traddr:xx.xxx.x.xx trsvcid:4420 subnqn:nqn.2016-06.io.spdk:cnode3'
EAL: DPDK is running on a NUMA system, but is compiled without NUMA support.
EAL: This will have adverse consequences for performance and usability.
EAL: Please use --legacy-mem option, or recompile with NUMA support.
TELEMETRY: No legacy callbacks, legacy socket not created
[2023-08-31 22:53:56.769533] nvme_fabric.c: 152:nvme_fabric_prop_get_cmd_sync: *ERROR*: Property Get failed
[2023-08-31 22:53:56.828530] nvme_fabric.c: 152:nvme_fabric_prop_get_cmd_sync: *ERROR*: Property Get failed
=====================================================
NVMe over Fabrics controller at xx.xxx.x.xx:4420: nqn.2016-06.io.spdk:cnode3
=====================================================
Controller Capabilities/Features
================================
Vendor ID:                             8086
Subsystem Vendor ID:                   8086
Serial Number:                         SPDK000000000003
Model Number:                          SPDK_Virtual_Controller_3
Firmware Version:                      23.09
Recommended Arb Burst:                 6
IEEE OUI Identifier:                   e4 d2 5c
Multi-path I/O
  May have multiple subsystem ports:   Yes
  May have multiple controllers:       Yes
  Associated with SR-IOV VF:           No
Max Data Transfer Size:                131072
Max Number of Namespaces:              32
Max Number of I/O Queues:              127
NVMe Specification Version (VS):       1.3
NVMe Specification Version (Identify): 1.3
Maximum Queue Entries:                 128
Contiguous Queues Required:            Yes
Arbitration Mechanisms Supported
  Weighted Round Robin:                Not Supported
  Vendor Specific:                     Not Supported
Reset Timeout:                         15000 ms
Doorbell Stride:                       4 bytes
NVM Subsystem Reset:                   Not Supported
Command Sets Supported
  NVM Command Set:                     Supported
Boot Partition:                        Not Supported
Memory Page Size Minimum:              4096 bytes
Memory Page Size Maximum:              4096 bytes
Persistent Memory Region:              Not Supported
Optional Asynchronous Events Supported
  Namespace Attribute Notices:         Supported
  Firmware Activation Notices:         Not Supported
  ANA Change Notices:                  Not Supported
  PLE Aggregate Log Change Notices:    Not Supported
  LBA Status Info Alert Notices:       Not Supported
  EGE Aggregate Log Change Notices:    Not Supported
  Normal NVM Subsystem Shutdown event: Not Supported
  Zone Descriptor Change Notices:      Not Supported
  Discovery Log Change Notices:        Not Supported
Controller Attributes
  128-bit Host Identifier:             Supported
  Non-Operational Permissive Mode:     Not Supported
  NVM Sets:                            Not Supported
  Read Recovery Levels:                Not Supported
  Endurance Groups:                    Not Supported
  Predictable Latency Mode:            Not Supported
  Traffic Based Keep ALive:            Not Supported
  Namespace Granularity:               Not Supported
  SQ Associations:                     Not Supported
  UUID List:                           Not Supported
  Multi-Domain Subsystem:              Not Supported
  Fixed Capacity Management:           Not Supported
  Variable Capacity Management:        Not Supported
  Delete Endurance Group:              Not Supported
  Delete NVM Set:                      Not Supported
  Extended LBA Formats Supported:      Not Supported
  Flexible Data Placement Supported:   Not Supported

Controller Memory Buffer Support
================================
Supported:                             No

Persistent Memory Region Support
================================
Supported:                             No

Admin Command Set Attributes
============================
Security Send/Receive:                 Not Supported
Format NVM:                            Not Supported
Firmware Activate/Download:            Not Supported
Namespace Management:                  Not Supported
Device Self-Test:                      Not Supported
Directives:                            Not Supported
NVMe-MI:                               Not Supported
Virtualization Management:             Not Supported
Doorbell Buffer Config:                Not Supported
Get LBA Status Capability:             Not Supported
Command & Feature Lockdown Capability: Not Supported
Abort Command Limit:                   4
Async Event Request Limit:             4
Number of Firmware Slots:              N/A
Firmware Slot 1 Read-Only:             N/A
Firmware Activation Without Reset:     N/A
Multiple Update Detection Support:     N/A
Firmware Update Granularity:           No Information Provided
Per-Namespace SMART Log:               No
Asymmetric Namespace Access Log Page:  Not Supported
Command Effects Log Page:              Supported
Get Log Page Extended Data:            Supported
Telemetry Log Pages:                   Not Supported
Persistent Event Log Pages:            Not Supported
Supported Log Pages Log Page:          May Support
Commands Supported & Effects Log Page: Not Supported
Feature Identifiers & Effects Log Page:May Support
NVMe-MI Commands & Effects Log Page:   May Support
Data Area 4 for Telemetry Log:         Not Supported
Error Log Page Entries Supported:      128
Keep Alive:                            Supported
Keep Alive Granularity:                10000 ms

NVM Command Set Attributes
==========================
Submission Queue Entry Size
  Max:                       64
  Min:                       64
Completion Queue Entry Size
  Max:                       16
  Min:                       16
Number of Namespaces:        32
Compare Command:             Supported
Write Uncorrectable Command: Not Supported
Dataset Management Command:  Supported
Write Zeroes Command:        Supported
Set Features Save Field:     Not Supported
Reservations:                Supported
Timestamp:                   Not Supported
Copy:                        Supported
Volatile Write Cache:        Present
Atomic Write Unit (Normal):  1
Atomic Write Unit (PFail):   1
Atomic Compare & Write Unit: 1
Fused Compare & Write:       Supported
Scatter-Gather List
  SGL Command Set:           Supported
  SGL Keyed:                 Supported
  SGL Bit Bucket Descriptor: Not Supported
  SGL Metadata Pointer:      Not Supported
  Oversized SGL:             Not Supported
  SGL Metadata Address:      Not Supported
  SGL Offset:                Supported
  Transport SGL Data Block:  Not Supported
Replay Protected Memory Block:  Not Supported

Firmware Slot Information
=========================
Active slot:                 1
Slot 1 Firmware Revision:    23.09


Commands Supported and Effects
==============================
Admin Commands
--------------
                  Get Log Page (02h): Supported
                      Identify (06h): Supported
                         Abort (08h): Supported
                  Set Features (09h): Supported
                  Get Features (0Ah): Supported
    Asynchronous Event Request (0Ch): Supported
                    Keep Alive (18h): Supported
I/O Commands
------------
                         Flush (00h): Supported LBA-Change
                         Write (01h): Supported LBA-Change
                          Read (02h): Supported
                       Compare (05h): Supported
                  Write Zeroes (08h): Supported LBA-Change
            Dataset Management (09h): Supported LBA-Change
                          Copy (19h): Supported LBA-Change
                       Unknown (79h): Supported LBA-Change
                       Unknown (7Ah): Supported

Error Log
=========

Arbitration
===========
Arbitration Burst:           1

Power Management
================
Number of Power States:          1
Current Power State:             Power State #0
Power State #0:
  Max Power:                      0.00 W
  Non-Operational State:         Operational
  Entry Latency:                 Not Reported
  Exit Latency:                  Not Reported
  Relative Read Throughput:      0
  Relative Read Latency:         0
  Relative Write Throughput:     0
  Relative Write Latency:        0
  Idle Power:                     Not Reported
  Active Power:                   Not Reported
Non-Operational Permissive Mode: Not Supported

Health Information
==================
Critical Warnings:
  Available Spare Space:     OK
  Temperature:               OK
  Device Reliability:        OK
  Read Only:                 No
  Volatile Memory Backup:    OK
Current Temperature:         0 Kelvin (-273 Celsius)
Temperature Threshold:       0 Kelvin (-273 Celsius)
Available Spare:             0%
Available Spare Threshold:   0%
Life Percentage Used:        0%
Data Units Read:             0
Data Units Written:          0
Host Read Commands:          0
Host Write Commands:         0
Controller Busy Time:        0 minutes
Power Cycles:                0
Power On Hours:              0 hours
Unsafe Shutdowns:            0
Unrecoverable Media Errors:  0
Lifetime Error Log Entries:  0
Warning Temperature Time:    0 minutes
Critical Temperature Time:   0 minutes

Number of Queues
================
Number of I/O Submission Queues:      127
Number of I/O Completion Queues:      127

Active Namespaces
=================
Namespace ID:1
Error Recovery Timeout:                Unlimited
Command Set Identifier:                NVM (00h)
Deallocate:                            Supported
Deallocated/Unwritten Error:           Not Supported
Deallocated Read Value:                Unknown
Deallocate in Write Zeroes:            Not Supported
Deallocated Guard Field:               0xFFFF
Flush:                                 Supported
Reservation:                           Supported
Namespace Sharing Capabilities:        Multiple Controllers
Size (in LBAs):                        2316435456 (8836GiB)
Capacity (in LBAs):                    2316435456 (8836GiB)
Utilization (in LBAs):                 2316435456 (8836GiB)
NGUID:                                 5BB3CC9E8D3E4EF2BE166FD5B831A1D0
UUID:                                  5bb3cc9e-8d3e-4ef2-be16-6fd5b831a1d0
Thin Provisioning:                     Not Supported
Per-NS Atomic Units:                   Yes
  Atomic Boundary Size (Normal):       0
  Atomic Boundary Size (PFail):        0
  Atomic Boundary Offset:              0
Maximum Single Source Range Length:    33
Maximum Copy Length:                   33
Maximum Source Range Count:            1
NGUID/EUI64 Never Reused:              No
Namespace Write Protected:             No
Number of LBA Formats:                 1
Current LBA Format:                    LBA Format #00
LBA Format #00: Data Size:  4096  Metadata Size:     0

**Comment by jimharris** [comment]
Can you try 127 for qpairs instead of 128 in the nvmf_create_transport RPC?  It is for number of IO qpairs, then when admin queue is added it is 128.

**Comment by sridharbalachandriah** [comment]
The above identify output was default nvmf_create_transport (without -m and -q)

**127 qpairs**

./scripts/rpc.py nvmf_create_transport -t tcp -u 131072 -i 131072  -m 127


nvme get-feature /dev/nvme0n1 -f 7 -H
get-feature:0x07 (Number of Queues), Current value:00000000
        Number of IO Completion Queues Allocated (NCQA): 1
        Number of IO Submission Queues Allocated (NSQA): 1


./build/examples/identify -r 'trtype:TCP adrfam:IPv4 traddr:xx.xxx.x.xx trsvcid:4420 subnqn:nqn.2016-06.io.spdk:cnode3'

Number of Queues

Number of I/O Submission Queues:      127
Number of I/O Completion Queues:      127


**32 qpairs**

./scripts/rpc.py nvmf_create_transport -t tcp -u 131072 -i 131072  -m 32

nvme get-feature /dev/nvme0n1 -f 7 -H
get-feature:0x07 (Number of Queues), Current value:00000000
        Number of IO Completion Queues Allocated (NCQA): 1
        Number of IO Submission Queues Allocated (NSQA): 1

./build/examples/identify -r 'trtype:TCP adrfam:IPv4 traddr:xx.xxx.x.xx trsvcid:4420 subnqn:nqn.2016-06.io.spdk:cnode3'

Number of Queues

Number of I/O Submission Queues:      32
Number of I/O Completion Queues:      32

**Comment by ksztyber** [comment]
That's weird, so our identify tool shows the expected values, but the kernel always reports 1. How many cores do you have on the initiator system? Also, did you have a chance to check this on a newer kernel?

**Comment by sridharbalachandriah** [comment]
The initiator system has 2 sockets with 32 cores each. I dint get a chance to have a new kernel on the system to isolate the issue. 

1. If the queue values are not getting set ( created), there is no easy way to get alerted. The dmesg output is also misleading

[468952.996109] nvme nvme0: new ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery", addr xx.xx.xx.xx:4420
[468952.996272] nvme nvme0: Removing ctrl: NQN "nqn.2014-08.org.nvmexpress.discovery"
[468953.005880] nvme nvme0: creating 127 I/O queues.
[468954.291011] nvme nvme0: mapped 127/0/0 default/read/poll queues.
[468954.322657] nvme nvme0: new ctrl: NQN "nqn.2016-06.io.spdk2:cnode2", addr xx.xx.xx.xx:442

3. Can one of you throw some light on the possible impact a single queue have for reaching the wire speed? As you see, there are enough cores for the nvmf_target and at the initiator end.

**Comment by ksztyber** [comment]
> [468953.005880] nvme nvme0: creating 127 I/O queues.
> [468954.291011] nvme nvme0: mapped 127/0/0 default/read/poll queues.

I think that should indicate that the driver is allocating 128 queues. Can you post the result of `scripts/rpc.py nvmf_get_stats` after `nvme connect`? Also, could you check if changing the `--nr-io-queues` parameter in `nvme connect` to something else, e.g 32, is changing anything?

> Can one of you throw some light on the possible impact a single queue have for reaching the wire speed? As you see, there are enough cores for the nvmf_target and at the initiator end.

The queues are mapped to cores, so if the kernel driver is really only creating a single queue, you'll be limited to the throughput of a single core.

**Comment by jimharris** [comment]
Can you also post the version of nvme-cli you are using?

I'm using name-cli v1.16, and I'm not able to reproduce this issue, even if I specify the namespace (/dev/nvme0n1) and not the controller (/dev/nvme0).

**Comment by krehm** [comment]
Just FYI, the problem is due in my opinion to a bug in the MOFED nvme-core driver.  I have opened a case
with Nvidia.   Bug exists in MOFED mlnx-nvme versions 5.7, 5.8. 5.9, 23.04, and I expect any other version
after 5.6.   We see it on RHEL 8.4 and Rocky 8.7.   Here is my post to Nvidia, they are just starting to look at the problem:

Starting with OFED 5.7-1.0.2.0 the following command fails:

[root@delphi-030 ~]# nvme get-feature /dev/nvme0n1 -f 7 -H
get-feature:0x07 (Number of Queues), Current value:00000000
Number of IO Completion Queues Allocated (NCQA): 1
Number of IO Submission Queues Allocated (NSQA): 1

The above output is incorrect, it should report 32 or 64 queues depending upon the NVME device. The problem is caused by a bug in function
nvme_submit_user_cmd() in file host/ioctl.c in mlnx-nvme. Here is the section
of code within that function from mlnx-nvme for OFED 5.6, which works:

nvme_execute_passthru_rq(req);
if (nvme_req(req)->flags & NVME_REQ_CANCELLED)
    ret = -EINTR;
else
    ret = nvme_req(req)->status;
if (result)
    *result = le64_to_cpu(nvme_req(req)->result.u64);
if (meta && !ret && !write) {
    if (copy_to_user(meta_buffer, meta, meta_len))
        ret = -EFAULT;
}

Here is the code which was introduced in 5.7, which is incorrect:

#if defined(HAVE_BLK_EXECUTE_RQ_2_PARAM) || defined(HAVE_BLK_EXECUTE_RQ_3_PARAM)
ret = nvme_execute_passthru_rq(req);
if (result)
    *result = le64_to_cpu(nvme_req(req)->result.u64);
#else
    nvme_execute_passthru_rq(req);
    if (nvme_req(req)->flags & NVME_REQ_CANCELLED)
        ret = -EINTR;
    else
        ret = nvme_req(req)->status;
#endif
if (meta && !ret && !write) {
    if (copy_to_user(meta_buffer, meta, meta_len))
        ret = -EFAULT;
}


What happens on our rocky-8.7 systems is that the #else clause is taken in the
above #if/#else/#endif clause. Note the code in the #if branch that
copies any non-zero result from the NVME device into *result, which in turn
means that that result will be copied back in the result field in the application's
ioctl request. If the #else branch is taken, this copy does not occur, and so the
*result field is left with the prior zero value in it. The user application gets
back a zero in *result when the NVMe device reported 32 or 64. Because the
queue field is zero-based, a zero *result gets translated to queue values of 1.

Note that any ioctl issued by an application in which the NVMe device returns
a non-zero result value will fail, the application will see a zero result.

The correct code should be:

#if defined(HAVE_BLK_EXECUTE_RQ_2_PARAM) || defined(HAVE_BLK_EXECUTE_RQ_3_PARAM)
    ret = nvme_execute_passthru_rq(req);
#else
    nvme_execute_passthru_rq(req);
    if (nvme_req(req)->flags & NVME_REQ_CANCELLED)
        ret = -EINTR;
    else
        ret = nvme_req(req)->status;
#endif
if (result)
    *result = le64_to_cpu(nvme_req(req)->result.u64);
if (meta && !ret && !write) {
    if (copy_to_user(meta_buffer, meta, meta_len))
        ret = -EFAULT;
}

**Comment by jimharris** [comment]
Thanks @krehm for the detailed response.  I've marked this as a 3rd party issue, but kept it open at least for now.  Please update this issue if and when you get resolution from Nvidia.

**Comment by krehm** [comment]
@jimharris I just heard back from Nvidia:  "I just wanted to inform you that the fix will be part of our MLNX_OFED 23.10 scheduled to be GA end of October."

**Comment by jimharris** [comment]
Thanks @krehm - based on your response I'm going to close this issue.

**Comment by krehm** [comment]
@jimharris The MOFED release with the fix is 23.10-0.5.5.0-LTS.  I have tested the fix on Rocky 8.4 and 8.7 and it works.
